src/components/views/rooms/RoomList.js

```diff
@@ -297,25 +298,40 @@ module.exports = React.createClass({
 
     // Update which rooms and users should appear according to which tags are selected
     updateVisibleRooms: function() {
-        this._visibleRooms = [];
-        FilterStore.getSelectedTags().forEach((tag) => {
-            (this._visibleRoomsForGroup[tag] || []).forEach(
-                (roomId) => this._visibleRooms.push(roomId),
+        const selectedTags = FilterStore.getSelectedTags();
+
+        let visibleGroupRooms = [];
+        selectedTags.forEach((tag) => {
+            visibleGroupRooms = visibleGroupRooms.concat(
+                this._visibleRoomsForGroup[tag] || [],
             );
         });
 
+        // If there are any tags selected, constrain the rooms listed to the
+        // visible rooms as determined by visibleGroupRooms. Here, we
+        // de-duplicate and filter out rooms that the client doesn't know
+        // about (hence the Set and the null-guard on `room`).
+        if (selectedTags.length > 0) {
+            const roomSet = new Set();
+            visibleGroupRooms.forEach((roomId) => {
+                const room = MatrixClientPeg.get().getRoom(roomId);
+                if (room) {
+                    roomSet.add(room);
+                }
+            });
+            this._visibleRooms = Array.from(roomSet);
+        } else {
+            // Show all rooms
+            this._visibleRooms = null;
```

Comment:
What I was trying to say was that you could also just do `this._visibleRooms = MatrixClientPeg.get().getRooms()` here (which would make this line of code less surprising in the context of the above comment).

Reply:
Oh I guess we can do the same when mounting. Would also make things better in terms of not using MatrixClientPeg when rendering.

Reply:
Actually we can't do that, because we need to get the total list of rooms on render otherwise we break some assumptions (when our client has new rooms added, we blindly `getRoomLists`).

Reply:
oh I see - would this not mean that newly added rooms wouldn't correctly appear if the list is filtered (where they otherwise should appear in that filter)?